### DISCUSSION TITLE
A simple `Slider` widget

### DISCUSSION THREAD

**Discussion by TomJGooding** [Textualize/textual#2824] [Show and tell]
I confess I'm a bit nervous to share this here, but I've been working on a simple `Slider` widget as an exercise to better understand Textual.

https://github.com/TomJGooding/textual-slider

Any feedback would be much appreciated!

**Comment by willmcgugan** [comment]
Looking forward to playing with it!

**Reply by TomJGooding** [reply]
Thanks Will - be gentle as I'm still getting to grips with the mouse/geometry in Textual!

**Comment by davep** [comment]
Looks really handy! Is this going onto PyPi?

**Reply by TomJGooding** [reply]
Thanks Dave! I really only created this as a learning exercise, I'm not sure it's ready to publish on PyPi yet but I might consider it!

I know that a built-in Slider widget is on the roadmap - but I'm assuming other widgets like the TextArea will be the priority?

**Reply by davep** [reply]
It deserves to be [on here](https://github.com/davep/transcendent-textual) but there's a good chance folk'll start asking for it via PyPi. Do it! DO IT!

> I know that a built-in Slider widget is on the roadmap - but I'm assuming other widgets like the TextArea will be the priority?

Eh. Alternative options aren't a bad thing. Heck, it's my job to work on Textual and I've implemented a third-party widget that'll eventually come to Textual but I wanted to have fun with it. Options are good!

**Reply by TomJGooding** [reply]
> ...there's a good chance folk'll start asking for it via PyPi. Do it! DO IT!

`textual-slider` is now on PyPi! Thanks Dave for your encouragement (I know who to blame when issues start coming in!)

**Reply by davep** [reply]
Awesome! Added it to [my list of stuff](https://github.com/davep/transcendent-textual).

**Reply by balthazarneveu** [reply]
This is a great project @TomJGooding  !

I'm currently writing a library called [interactive_pipe](https://github.com/balthazarneveu/interactive_pipe) which is (in summary) an image/signal processing framework with sliders...

Been maturing this one since 2020 , used it in industry . Helped to make many algorithms much easier to debug & explain, including visualizing how neural networks react when we continuously change image properties. 

Some screenshots to get the idea as I'm rewriting the whole core right now. 
![image](https://github.com/Textualize/textual/assets/41070742/d9966fb4-09ac-4e7b-8fcc-c6745869b61b)
It supports PyQT  (& soon Matplotlib & Jupyter IPyWidgets backends)

Since `interactive_pipe` can do many other stuffs (such as a jukebox for my baby daughter on the raspberry pi) as the syntax is pretty elegant, I'm thinking about adding a **Textual backend.** . 
Probably not going to explore fully displaying images as ascii art, but I think of it more like a good way to run the processing engine by being able to play with sliders & without needing heavy GUI stuff such as PyQt or matplotlib . 

and since sliders are a mandatory piece for that framework, here's my appreciation post :heart: 

![image](https://github.com/Textualize/textual/assets/41070742/b5c8941d-4790-4455-8ed6-dcc18f5f92ed)


:heart:  Thanks a lot & generally speaking to the textual developpers , just got that "wow" effect when I discovered it :+1: :+1:

**Reply by TomJGooding** [reply]
Thanks @balthazarneveu, glad this might be useful and it sounds like an interesting project!

But really the Textual devs deserve the credit, my slider widget just builds upon the existing implementation of a scrollbar.

**Comment by OxFF00FF** [comment]
How to create range slider for choose time values 00:00:10 and 12:47:00 ?

00:00:10  ────●━━━━●──────  12:47:00
------------------↑ --------- ↑
---------------- Start ----- End
                      


reference 
<img width="634" height="212" alt="image" src="https://github.com/user-attachments/assets/9334c888-f1c6-43f7-8800-899ed3ed34e1" />

**Reply by TomJGooding** [reply]
This probably wouldn't give the fine-grained control you're looking for, since terminals work in terms of character cells rather than pixels. 

For example, assuming a slider around 50 characters wide and a 24 hour time period, this would only allow 30-minute intervals:

```
H 0000000000000000000011111111111111111111222222222
H 0011223344556677889900112233445566778899001122334
  :::::::::::::::::::::::::::::::::::::::::::::::::
M 0303030303030303030303030303030303030303030303030
M 0000000000000000000000000000000000000000000000000

  ●━━━━━━━━━━━━━━━━━━━━━━━━━●──────────────────────
  ↑                         ↑
  00:00:00                  13:00:00
```

**Reply by OxFF00FF** [reply]
I was just giving an example. for my project, I won't need a duration longer than an hour. I’m curious whether it’s technically possible to create such a widget.

**Reply by TomJGooding** [reply]
It's _technically_ possible to create all sorts of custom widgets in Textual - for example when I was learning Textual I also created a [color picker](https://github.com/TomJGooding/textual-colorpicker). I'd suggest starting with the docs about building [custom widgets](https://textual.textualize.io/guide/widgets/#custom-widgets).
